ably/realtime_channel.go

```diff
@@ -2,130 +2,342 @@ package ably
+	return c.PublishAll([]*proto.Message{{Name: name, Data: data}}, listen)
+}
+
+// PublishAll publishes all given messages on the channel at once.
+// PublishAll does not block.
+//
+// If listen channel is non-nil it will be used to notify whether the messages
+// were published succesfully with nil error value; when the operation failed,
+// the error that caused the failure will be sent on the listen channel.
+//
+// This implicitly attaches the channel if it's not already attached.
+func (c *RealtimeChannel) PublishAll(messages []*proto.Message, listen chan<- error) error {
+	if err := c.Attach(); err != nil {
+		return err
+	}
 	msg := &proto.ProtocolMessage{
-		Action:  proto.ActionMessage,
-		Channel: c.Name,
-		Messages: []*proto.Message{
-			{Name: name, Data: data},
-		},
+		Action:   proto.ActionMessage,
+		Channel:  c.Name,
+		Messages: messages,
 	}
-	return c.client.send(msg)
+	switch c.State() {
+	case StateChanInitialized, StateChanAttaching:
+		if c.client.opts.NoQueueing {
+			return errQueueing
+		}
+		c.queue.Enqueue(msg, listen)
+		return nil
+	case StateChanAttached:
+	default:
+		return &Error{Code: 90001}
+	}
+	return c.client.Connection.send(msg, listen)
+}
+
+// State gives current state of the channel.
+func (c *RealtimeChannel) State() int {
+	c.state.Lock()
+	state := c.state.current
+	c.state.Unlock()
+	return state
+}
+
+// Reason gives the last error that caused channel transition to failed state.
+func (c *RealtimeChannel) Reason() error {
+	c.state.Lock()
+	err := c.state.err
+	c.state.Unlock()
+	return err
 }
 
 func (c *RealtimeChannel) notify(msg *proto.ProtocolMessage) {
 	switch msg.Action {
 	case proto.ActionAttached:
-		c.setState(ChanStateAttached)
+		c.state.syncSet(StateChanAttached, nil)
+		c.queue.Flush()
 	case proto.ActionDetached:
-		c.setState(ChanStateDetached)
+		c.state.syncSet(StateChanDetached, nil)
 	case proto.ActionPresence:
-		// TODO
+		Log.Printf(LogError, "TODO: implement RealtimePresence: %v", msg)
 	case proto.ActionError:
-		c.setState(ChanStateFailed)
-		c.Err = msg.Error
-		// TODO c.Close()
+		c.state.syncSet(StateChanFailed, newErrorProto(msg.Error))
+		c.queue.Fail(newErrorProto(msg.Error))
+		// TODO recovery
 	case proto.ActionMessage:
-		c.listenMtx.RLock()
-		defer c.listenMtx.RUnlock()
-		for _, m := range msg.Messages {
-			if l, ok := c.listeners[""]; ok {
+		c.listenMtx.Lock()
+		for _, msg := range msg.Messages {
+			if l, ok := c.listen[""]; ok {
 				for ch := range l {
-					ch <- m
+					select {
+					case ch <- msg:
+					default:
+						Log.Printf(LogWarn, "dropped %q message due to slow receiver", msg.Name)
+					}
 				}
 			}
-			if l, ok := c.listeners[m.Name]; ok {
+			if l, ok := c.listen[msg.Name]; ok {
 				for ch := range l {
-					ch <- m
+					select {
+					case ch <- msg:
+					default:
+						Log.Printf(LogWarn, "dropped %q message due to slow receiver", msg.Name)
+					}
 				}
 			}
```

Comment:
Rather than storing channels, I would store a wrapper type which is responsible for sending the messages to the subscriber, e.g.:

```go
type Subscription struct {
	ch    chan *proto.Message
	queue Queue
}

func newSubscription(ch chan *proto.Message) *Subscription {
	s := &Subscription{
		ch:    ch,
		queue: newQueue(),
	}
	go s.loop()
	return s
}

func (s *Subscription) notify(msg *proto.Message) {
	s.queue.Enqueue(msg)
}

func (s *Subscription) loop() {
	for {
		s.ch <- s.queue.Pop()
	}
}
```

then you don't have to worry about slow receivers, and also allows the user to just call something like `subscription.Close()` to unsubscribe / close the channel.